cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -2881,22 +2881,61 @@ pub(crate) fn emit(
             let i2 = Inst::mov_r_r(OperandSize::Size64, rax, r11_w);
             i2.emit(sink, info, state);
 
-            // opq %r10, %r11
             let r10_rmi = RegMemImm::reg(r10);
-            let i3 = if *op == inst_common::AtomicRmwOp::Xchg {
-                Inst::mov_r_r(OperandSize::Size64, r10, r11_w)
-            } else {
-                let alu_op = match op {
-                    inst_common::AtomicRmwOp::Add => AluRmiROpcode::Add,
-                    inst_common::AtomicRmwOp::Sub => AluRmiROpcode::Sub,
-                    inst_common::AtomicRmwOp::And => AluRmiROpcode::And,
-                    inst_common::AtomicRmwOp::Or => AluRmiROpcode::Or,
-                    inst_common::AtomicRmwOp::Xor => AluRmiROpcode::Xor,
-                    inst_common::AtomicRmwOp::Xchg => unreachable!(),
-                };
-                Inst::alu_rmi_r(OperandSize::Size64, alu_op, r10_rmi, r11_w)
-            };
-            i3.emit(sink, info, state);
+            match op {
+                inst_common::AtomicRmwOp::Xchg => {
```

Comment:
This could be optimized to just the `xchg` instruction without an `cmpxchg` loop, but that could be done at a later time.